### DISCUSSION TITLE
Optimize me #4, Sounds of abyss geysers

### DISCUSSION THREAD

**Discussion by SomeRandomNoobKekeke** [FakeFishGames/Barotrauma#15236] [Ideas and suggestions]
Updating the sounds seems to be the heaviest part of LevelObject.Update, at least for geysers

E.g. on ridge's abyss mountain update of "blacksmoker1" sounds could take up to 1ms

I mean this:
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaClient/ClientSource/Map/Levels/LevelObjects/LevelObject.cs#L225-L248
![20241123062933_1](https://github.com/user-attachments/assets/06c5dec5-bb55-4d50-be0c-950fbabfdc37)

almost entirely this call
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaClient/ClientSource/Map/Levels/LevelObjects/LevelObject.cs#L238

![20241123070000_1](https://github.com/user-attachments/assets/fe10225a-0fca-4228-9056-c383137a7098)

Also it seems that they're not just constantly lagging, they are also spiking

Here's 144p video of how exactly they are updated. That's not real fps, i set capture duration to only 1 frame and added fakelags to slow updates down

https://github.com/user-attachments/assets/d6418429-8437-4b95-945b-dd108b0f97d8

**Comment by Regalis11** [comment]
I gave this a test and did some measurements with visual studio's profiler, but I couldn't spot any performance issues here. In a spot with about a dozen hydrothermal vents visible and playing sounds, the total CPU usage of all `LevelObject.Update` calls was only 0.04%:

![image](https://github.com/user-attachments/assets/4d6716b7-855e-4cb5-8efd-99ed90239734)

So I suspect that his either requires some more specific circumstances to occur, or that there's something wrong in your measurements. What does this custom debug view actually measure, what does that reading mean and how does it do the measurement?

**Reply by SomeRandomNoobKekeke** [reply]
Well, idk, there's just stopwatch calls around that, litteraly this:

https://github.com/SomeRandomNoobKekeke/Showperf-extensions--code-/blob/main/CSharp/Client/Patches/Capture/Map/Levels/LevelObjects/LevelObject.cs#L157-L196

I was testing with no other mods but in luatrauma, it could be just bug of luatrauma, or mb some specific audio settings, or mb it happens only on potato PCs

I'll try in full vanilla, with different settings, but after that i'm totally clueless
![barodev small](https://github.com/user-attachments/assets/0f91c060-64a7-4b08-b7c2-383dad72d39c)

**Reply by SomeRandomNoobKekeke** [reply]
No, it's not caused by luatrauma, and not by settings
Must be the potatoesness of PC
~Mb you could add something like Thread.Sleep in gamescreen update to simulate potatoesness?~
~And mb a command to set those fakelags?~
~Something like zis: https://github.com/SomeRandomNoobKekeke/Showperf-extensions--code-/blob/main/CSharp/Client/Commands.cs#L99-L113~
Probably won't help

**Reply by SomeRandomNoobKekeke** [reply]
You probably just unlucky, it seems to not always happen on normal levels (mb only when they sync somehow?)
Try here https://steamcommunity.com/sharedfiles/filedetails/?id=3371681139

Also it seems that total Sound.Length of level objects near abyss mountain is the highest (which is natural), ~200 on lagridge
Could it be related?

And also it seems that "roundSound.Sound.Play" doesn't stick out for most of the level, only near geysers where Sound.Length is high
![20241125072054_1](https://github.com/user-attachments/assets/fac99b7d-4d3d-4081-aed8-2d8086d2d4ae)

**Reply by Regalis11** [reply]
> Also it seems that total Sound.Length of level objects near abyss mountain is the highest (which is natural), ~200 on lagridge
Could it be related?

To me it sounds like this must be an issue in the way you're doing the measurements: `Sounds.Length` is just a getter for the length of the array.

**Reply by SomeRandomNoobKekeke** [reply]
Guh, yes, it's array length, that's not ms, it's count
I mean the biggest concentration of some LevelObjects with some sounds is in the abyss, up to 200 sounds simultaneously attempted to be played by level objects
Iirc 2 per lavaevent and 1 per blacksmoker1
Idk how it works internally and could this be a problem, so i'm asking
![barodev small](https://github.com/user-attachments/assets/40abd463-0dea-4fce-b1f3-c3823988bee0)

I assume they sync and overload some OpenAL mechanism, thats would explain why lags are inconsistent with medium number of geysers like a dozen, and guaranteed to happen on large numbers

And i can see small optimization here: Why does lavaevent have 2 sounds: Thunder1.ogg and BlackSmoker.ogg
If it's errupted from a blacksmoker which also plays BlackSmoker.ogg?
Why would you play that twice?
![BaroDev (wide)](https://github.com/user-attachments/assets/fd4f60c7-5001-4da6-9239-40d481fe579c)

> To me it sounds like this must be an issue in the way you're doing the measurements

I'm sure there's no issue with the measurements: i can see same lags with vanilla showperf and if i comment out that code levelobject sounds stops updating and lags passes
![BaroDev (ultra wide)](https://github.com/user-attachments/assets/016e820c-5812-481d-935b-0a9259f0267c)

**Reply by Regalis11** [reply]
Hmm, I guess I was confused about the label `Sounds.Length` here: I assumed the measurement meant the `Sounds.Length` call (and it would be really strange for simply accessing the array's length to take a noticeable amount of time), but it seems that it actually measures everything inside the for loop that goes through the sounds.

But do I understand this correctly, that number is the amount of ticks? If that's the case, 106 is nothing, that's 0.0106 milliseconds.

**Reply by SomeRandomNoobKekeke** [reply]
No, number of elements in Sounds arrays, i just can capture any number so i just putted number of objects there instead of ticks
![barodev small](https://github.com/user-attachments/assets/c6cab11c-6f28-42af-baf0-d272b226d4af)

**Comment by SomeRandomNoobKekeke** [comment]
Wait, did you mean that i count number of sounds of updated LevelObjects but ignoring the fact that most will get culled from listener pos and won't be played at once?
That's fair, i captured this because i was just curious how many there are, theoretically
Lets capture the number of roundSound.Sound.Play calls from level objects per frame instead :baromodder:
![20241126095807_1](https://github.com/user-attachments/assets/198f3c0b-2ab2-458a-a777-0a64528e0cea)
![20241126095816_1](https://github.com/user-attachments/assets/8b728915-3ed2-4ffd-be6e-36cdef89d0ce)
To be fair, it's hard to find a spot with more than 12 calls at once in ridgeadvanced, but still, that's the largest amount on the map, and idk how many similar calls are made from other entities at the same moment

It may not be the cause, it's just funny coincidence that number of calls correlates with presence of roundSound.Sound.Play spikes
